crates/storage/provider/src/providers/rocksdb/invariants.rs

```diff
@@ -0,0 +1,688 @@
+    ) -> ProviderResult<Option<BlockNumber>>
+    where
+        Provider: DBProvider + StageCheckpointReader,
+    {
+        // Get the IndexStorageHistory stage checkpoint
+        let checkpoint = provider
+            .get_stage_checkpoint(StageId::IndexStorageHistory)?
+            .map(|cp| cp.block_number)
+            .unwrap_or(0);
+
+        // Check if RocksDB has any data
+        let rocks_first = self.first::<tables::StoragesHistory>()?;
+
+        match rocks_first {
+            Some(_) => {
+                // If checkpoint is 0 but we have data, clear everything
+                if checkpoint == 0 {
+                    tracing::info!(
+                        target: "reth::providers::rocksdb",
+                        "StoragesHistory has data but checkpoint is 0, clearing all"
+                    );
+                    self.prune_storages_history_above(0)?;
+                    return Ok(None);
+                }
+
+                // Find the max highest_block_number (excluding u64::MAX sentinel) across all
+                // entries
+                let mut max_highest_block = 0u64;
+                for result in self.iter::<tables::StoragesHistory>()? {
+                    let (key, _) = result?;
+                    let highest = key.sharded_key.highest_block_number;
+                    if highest != u64::MAX && highest > max_highest_block {
+                        max_highest_block = highest;
+                    }
+                }
+
+                // If any entry has highest_block > checkpoint, prune excess
+                if max_highest_block > checkpoint {
+                    tracing::info!(
+                        target: "reth::providers::rocksdb",
+                        rocks_highest = max_highest_block,
+                        checkpoint,
+                        "StoragesHistory ahead of checkpoint, pruning excess data"
+                    );
+                    self.prune_storages_history_above(checkpoint)?;
+                }
+
+                Ok(None)
+            }
+            None => {
+                // Empty RocksDB table
+                if checkpoint > 0 {
+                    // Stage says we should have data but we don't
+                    return Ok(Some(0));
+                }
+                Ok(None)
+            }
+        }
+    }
+
+    /// Prunes `StoragesHistory` entries where `highest_block_number` > `max_block`.
+    ///
+    /// For `StoragesHistory`, the key contains `highest_block_number`, so we can iterate
+    /// and delete entries where `key.sharded_key.highest_block_number > max_block`.
+    fn prune_storages_history_above(&self, max_block: BlockNumber) -> ProviderResult<()> {
+        use reth_db_api::models::storage_sharded_key::StorageShardedKey;
+
+        let mut to_delete: Vec<StorageShardedKey> = Vec::new();
+
+        // Iterate all entries and collect those to delete
+        for result in self.iter::<tables::StoragesHistory>()? {
```

Comment:
similarly we can't iterate like this. `crates/prune/prune/src/segments/user/storage_history.rs` we should try to emulate what pruner does and/or aggregate the pruning logic

Reply:
but, we actually need storage changesets in static files first for this.

the way entries are selected are by looking at storage changesets for the block range. however, since StorageChangeSets are still in DB, and it is behind alongside the checkpoint, we cant really know which kets we need to compose to prune here (apart from iterating like this)

context:
https://github.com/paradigmxyz/reth/blob/948966781425b6ee6d9fec7e2a1078d58eacec04/crates/prune/prune/src/segments/user/storage_history.rs#L82-L113